### ISSUE TITLE
System Level Performance Engineering

### ISSUE THREAD

**Issue by nirmalpratheep** [The-School-of-AI/LLM#223]
# Enhancement: System-Level Performance Profiling for DeepSpeed MoE Configurations

## 🎯 Overview

Execute standardized system-level performance profiling for DeepSpeed MoE (Mixture of Experts) training to identify bottlenecks, validate stability, and ensure optimal resource utilization across ZeRO-2 and ZeRO-3 stages.

## 🔬 Target Commands

```bash
# Test ZeRO-2 MoE config
deepspeed --num_gpus=4 __main.py__ --deepspeed_config config/deepspeed/zero-2-moe.json

# Test ZeRO-3 MoE config
deepspeed --num_gpus=4 __main.py__ --deepspeed_config config/deepspeed/zero-3-moe.json
```

---

### Step 1: Enable Runtime Telemetry

**Environment Variables:**
```bash
export NCCL_DEBUG=WARN
export TORCH_DISTRIBUTED_DEBUG=DETAIL
export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
```

**DeepSpeed Configuration:**
Add to both `zero-2-moe.json` and `zero-3-moe.json`:
```json
{
  "wall_clock_breakdown": true
}
```

**Tasks:**
- [ ] Set environment variables in launch script
- [ ] Update DeepSpeed config files
- [ ] Verify telemetry logging in output

---

### Step 2: GPU Utilization Monitoring

**Command:**
```bash
nvidia-smi dmon -s pucvmt
```

Run in separate terminal during training to capture:
- **SM utilization** (Streaming Multiprocessor usage)
- **Memory utilization** (GPU memory active usage)
- **Power draw** (watts consumed)
- **Temperature** (thermal state)
- **Idle gaps** (periods of GPU inactivity)

**Acceptance Criteria:**
- ✅ GPU utilization ≥80% sustained (post-warmup)
- 🚨 **Red Flag:** GPU utilization <80% sustained

**Tasks:**
- [ ] Start `nvidia-smi dmon` before training
- [ ] Log output to file: `nvidia-smi dmon -s pucvmt > gpu_util.log`
- [ ] Analyze post-warmup steady state (skip first 50-100 steps)
- [ ] Document any sustained drops below 80%

---

### Step 3: Throughput Stability Tracking

**Metrics to Log (Per Step):**
- Step duration (seconds)
- Tokens/sec
- Samples/sec

**Acceptance Criteria:**
- ✅ Post-warmup variance ≤ ±5%
- 🚨 **Red Flag:** Tokens/sec drops >5% mid-run

**Tasks:**
- [ ] Instrument training loop to log step metrics
- [ ] Calculate mean and standard deviation after warmup period
- [ ] Plot tokens/sec over time
- [ ] Identify any degradation trends

**Example Logging Code:**
```python
import time
step_start = time.time()
# ... training step ...
step_duration = time.time() - step_start
tokens_per_sec = (batch_size * seq_length * world_size) / step_duration
print(f"Step {step}, Duration: {step_duration:.3f}s, Tokens/s: {tokens_per_sec:.0f}")
```

---

### Step 4: Communication vs Compute Breakdown

**Using DeepSpeed Wall-Clock Timers:**

DeepSpeed automatically captures timing when `"wall_clock_breakdown": true` is enabled.

**Metrics to Extract:**
- Forward compute time
- Backward compute time
- Gradient reduce time (all-reduce)
- Parameter gather/scatter time (ZeRO-3 specific)

**Goal:**
- Communication should be **overlapped** with compute
- Communication overhead <40% of total step time

**Acceptance Criteria:**
- ✅ Communication overhead ≤40% of step time
- 🚨 **Red Flag:** Communication exceeds 40% of step time

**Tasks:**
- [ ] Parse DeepSpeed logs for wall-clock breakdown
- [ ] Calculate communication percentage: `(reduce_time + gather_time) / step_time * 100`
- [ ] Verify gradient communication overlaps with backward pass
- [ ] Compare ZeRO-2 vs ZeRO-3 communication patterns

**Example Output:**
```
[DeepSpeed] step=100, forward=120ms, backward=180ms, reduce=40ms, total=340ms
Communication %: 11.8%
```

---

### Step 5: CPU & Data Pipeline Health

**Monitoring Commands:**
```bash
# CPU utilization
top -b -n 1 | head -20
# or
htop

# I/O statistics
iostat -xm 1
```

**Metrics to Track:**
- CPU utilization per core
- I/O wait percentage
- Data loader queue depth
- Disk read/write throughput

**Acceptance Criteria:**
- ✅ CPU utilization <90% (headroom for system tasks)
- ✅ I/O wait <10%
- 🚨 **Red Flag:** CPU >90% or I/O wait >10%

**Tasks:**
- [ ] Monitor CPU during training
- [ ] Check for CPU-bound data preprocessing
- [ ] Verify data loader workers are sufficient
- [ ] Ensure storage throughput meets demand
- [ ] Profile data loading pipeline if bottleneck detected

**Common Issues:**
- Insufficient `num_workers` in DataLoader
- CPU-heavy tokenization/augmentation
- Slow storage (network mounts, slow disks)

---

### Step 6: Memory Efficiency Validation

**Tracking Metrics:**
- Allocated vs reserved GPU memory
- Memory fragmentation behavior
- ZeRO-3 partition stability
- Progressive memory growth

**Monitoring:**
```bash
# Continuous memory monitoring
nvidia-smi --query-gpu=memory.used,memory.free --format=csv -l 1 > memory.log

# Or use PyTorch profiling
torch.cuda.memory_summary()
```

**Acceptance Criteria:**
- ✅ No progressive memory growth over time
- ✅ No fragmentation-induced stalls or OOM errors
- ✅ Stable memory footprint after warmup

**Tasks:**
- [ ] Log memory usage throughout training
- [ ] Check for memory leaks (gradual increase)
- [ ] Validate ZeRO-3 partition consistency
- [ ] Monitor fragmentation: `torch.cuda.memory_stats()`
- [ ] Document peak memory usage for both ZeRO-2 and ZeRO-3

**ZeRO-3 Specific:**
- Monitor parameter gathering/partitioning cycles
- Ensure no memory spikes during optimizer steps

---

### Step 7: Network Utilization

**Monitoring Commands:**

For NVLink:
```bash
nvidia-smi nvlink --status
# or
nvidia-smi nvlink -g 0 -l 1  # live monitoring GPU 0
```

For InfiniBand/Ethernet:
```bash
ifstat
# or
ibstat
# or
sar -n DEV 1
```

**Metrics to Track:**
- Bandwidth utilization (GB/s)
- Link imbalance (uneven traffic across links)
- Saturation events (dropped packets, congestion)
- Link errors

**Acceptance Criteria:**
- ✅ Bandwidth utilization aligned with communication volume
- ✅ Balanced traffic across all links/GPUs
- ✅ No link errors or saturation

**Tasks:**
- [ ] Monitor NVLink bandwidth during training
- [ ] Check for asymmetric communication patterns
- [ ] Verify all GPUs communicate efficiently
- [ ] Document peak bandwidth utilization
- [ ] Identify any network bottlenecks

---

### Step 8: Spot Interruption Safety (Pause/Resume)

**Checkpoint Save/Resume Validation:**

Simulate SIGTERM (spot instance interruption):
```bash
# During training, send SIGTERM
kill -15 <training_process_pid>
```

**Validation Steps:**
1. **Checkpoint Save Latency**
   - Measure time from SIGTERM to checkpoint completion
   - Target: <60 seconds for clean shutdown

2. **Clean Shutdown**
   - Verify checkpoint integrity
   - No corrupted files
   - All GPU resources released

3. **Resume Training**
   - Restart from checkpoint (local or S3)
   - Verify training continues without errors

**Acceptance Criteria:**
- ✅ Checkpoint saves successfully within timeout window
- ✅ Resume without throughput regression (within ±2%)
- ✅ No accuracy/loss divergence after resume

**Tasks:**
- [ ] Test checkpoint save on SIGTERM
- [ ] Verify checkpoint upload to S3 (if applicable)
- [ ] Resume training from checkpoint
- [ ] Compare pre-interrupt vs post-resume throughput
- [ ] Validate loss curve continuity
- [ ] Test multiple interrupt/resume cycles

**Example Test:**
```bash
# Terminal 1: Start training
deepspeed --num_gpus=4 __main.py__ --deepspeed_config config/deepspeed/zero-3-moe.json

# Terminal 2: After 100 steps, simulate interruption
kill -15 $(pgrep -f deepspeed)

# Terminal 1: Resume from checkpoint
deepspeed --num_gpus=4 __main.py__ --deepspeed_config config/deepspeed/zero-3-moe.json --resume_from_checkpoint ./checkpoints/step_100
```

---

## 🚨 Escalation Triggers

**Immediate escalation required if ANY of the following occur:**

1. **Throughput Degradation**
   - Tokens/sec drops >5% mid-run (outside warmup)

2. **GPU Underutilization**
   - GPU utilization <75% sustained (>100 consecutive steps)

3. **Communication Overhead**
   - Communication exceeds 40% of step time

4. **CPU/I/O Stalls**
   - CPU >90% sustained
   - I/O wait >10% sustained
   - GPUs idle waiting for data

5. **Resume Regression**
   - Resume from checkpoint causes step-time regression >2%

6. **Memory Issues**
   - Progressive memory growth (leak)
   - Fragmentation-induced OOM
   - Unexpected memory spikes

**Escalation Actions:**
- [ ] Stop current profiling run
- [ ] Capture full diagnostic snapshot
- [ ] Document exact conditions when issue occurred
- [ ] Notify performance engineering team
- [ ] Preserve logs and checkpoints for analysis

---

### 1. Profiling Report
- [ ] System-level metrics summary (all 8 steps)
- [ ] ZeRO-2 vs ZeRO-3 comparison table
- [ ] Bottleneck identification with severity ranking
- [ ] Throughput stability analysis with variance metrics

### 2. Log Artifacts
- [ ] `gpu_util.log` - GPU utilization timeline
- [ ] `memory.log` - Memory usage over time
- [ ] DeepSpeed wall-clock breakdown logs
- [ ] CPU/I/O monitoring logs
- [ ] Network bandwidth logs
- [ ] Checkpoint save/resume test results

### 3. Performance Summary
**Required Metrics:**
- Steady-state tokens/sec (mean ± std)
- GPU utilization % (mean, min, max)
- Communication overhead %
- Memory efficiency (allocated/total)
- Checkpoint save latency
- Resume throughput delta

### 4. Go/No-Go Recommendation
Based on acceptance criteria, provide:
- ✅ **GO** - All criteria met, ready for production
- ⚠️ **CONDITIONAL** - Minor issues, requires tuning
- 🚨 **NO-GO** - Critical issues, requires investigation

---

### Hardware Configuration
```
GPU Type: _____________
GPU Memory: _____________
GPU Count: 4
Interconnect: _____________ (NVLink/InfiniBand/PCIe)
CPU: _____________
RAM: _____________
Storage: _____________ (local SSD / network / S3)
```

### Software Stack
```
DeepSpeed Version: _____________
PyTorch Version: _____________
CUDA Version: _____________
NCCL Version: _____________
Python Version: _____________
```

### Model Configuration
```
Model Size: _____________ parameters
Number of Experts: _____________
Expert Capacity Factor: _____________
Sequence Length: _____________
Batch Size (per GPU): _____________
Gradient Accumulation Steps: _____________
```

---

**Labels:** `enhancement`, `performance`, `profiling`, `deepspeed`, `system-level`, `moe`

**Priority:** High

**Assignees:** @performance-team

**Milestone:** System Profiling Sprint
